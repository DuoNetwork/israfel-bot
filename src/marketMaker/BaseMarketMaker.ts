import {
	Constants as WrapperConstants,
	DualClassWrapper,
	IDualClassStates,
	VivaldiWrapper,
	Web3Wrapper
} from '@finbook/duo-contract-wrapper';
import { Constants as DataConstants, IPrice } from '@finbook/duo-market-data';
import {
	Constants,
	IAccount,
	// IOrderBookSnapshot,
	IOrderBookSnapshotLevel,
	IToken,
	IUserOrder,
	// OrderBookUtil,
	RelayerClient,
	Util,
	Web3Util
} from '@finbook/israfel-common';
import * as CST from '../common/constants';
import { IOption } from '../common/types';
import util from '../utils/util';

export default class BaseMarketMaker {
	public tokens: IToken[] = [];
	public liveBidOrders: IUserOrder[][] = [[], []];
	public liveAskOrders: IUserOrder[][] = [[], []];
	public makerAccount: IAccount = { address: '0x0', privateKey: '' };
	public custodianStates: IDualClassStates | null = null;
	public priceStep: number = 0.0001;
	public tokenBalances: number[] = [0, 0, 0];
	public pendingOrders: { [orderHash: string]: boolean } = {};
	public exchangePrices: { [source: string]: IPrice[] } = {};
	public isBeethoven = true;
	public isInitialized = false;
	public isSendingOrder = false;
	public isMaintainingBalance = false;
	public isMakingOrders = false;
	public lastMid: number[] = [0, 0];

	public isA(pair: string) {
		return pair.startsWith(this.tokens[0].code);
	}

	public getOtherPair(pair: string) {
		const isA = this.isA(pair);
		return this.tokens[isA ? 1 : 0].code + '|' + pair.split('|')[1];
	}

	public getEthPrice() {
		return this.exchangePrices[DataConstants.API_KRAKEN] &&
			this.exchangePrices[DataConstants.API_KRAKEN].length
			? this.exchangePrices[DataConstants.API_KRAKEN][0].close
			: 0;
	}

	public async checkAllowance(web3Util: Web3Util, duoWrapper: DualClassWrapper | VivaldiWrapper) {
		Util.logDebug('start to check allowance');
		const address = this.makerAccount.address;
		for (const code of [Constants.TOKEN_WETH, this.tokens[0].code, this.tokens[1].code])
			if (!(await web3Util.getTokenAllowance(code, address))) {
				Util.logDebug(`${address} ${code} allowance is 0, approving`);
				const txHash = await web3Util.setUnlimitedTokenAllowance(code, address);
				await web3Util.awaitTransactionSuccessAsync(txHash);
			}

		const custodianAddress = duoWrapper.address;
		if (!(await web3Util.getTokenAllowance(Constants.TOKEN_WETH, address, custodianAddress))) {
			Util.logDebug(`${address} for custodian allowance is 0, approving`);
			const txHash = await web3Util.setUnlimitedTokenAllowance(
				Constants.TOKEN_WETH,
				address,
				custodianAddress
			);
			await web3Util.awaitTransactionSuccessAsync(txHash);
		}
		Util.logDebug('completed checking allowance');
	}

	public async maintainBalance(
		web3Util: Web3Util,
		duoWrapper: DualClassWrapper | VivaldiWrapper
	) {
		util.logDebug(
			`obj address web3: ${web3Util.valueOf()} duoWrapper: ${duoWrapper.valueOf()}`
		);
		throw new Error('not implemented method');
	}

	public canMakeOrder(relayerClient: RelayerClient, pair: string) {
		const otherPair = this.getOtherPair(pair);
		if (
			!relayerClient.orderBookSnapshots[pair] ||
			!relayerClient.orderBookSnapshots[otherPair]
		) {
			Util.logDebug('waiting for the other orderbook');
			return false;
		}

		if (this.isSendingOrder || !Util.isEmptyObject(this.pendingOrders)) {
			Util.logDebug(`non empty pending updates ${Object.keys(this.pendingOrders)}`);
			return false;
		}

		return true;
	}

	public async takeOneSideOrders(
		relayerClient: RelayerClient,
		pair: string,
		isBid: boolean,
		orderBookSide: IOrderBookSnapshotLevel[]
	) {
		this.isSendingOrder = true;
		for (const orderLevel of orderBookSide) {
			Util.logDebug(
				`${pair} taking an ${isBid ? 'bid' : 'ask'} order with price ${
					orderLevel.price
				} amount ${orderLevel.balance}`
			);
			if (!orderLevel.balance) continue;
			const orderHash = await relayerClient.addOrder(
				this.makerAccount.address,
				pair,
				orderLevel.price,
				orderLevel.balance,
				!isBid,
				Util.getExpiryTimestamp(true)
			);
			this.pendingOrders[orderHash] = true;
			await Util.sleep(1000);
		}
		this.isSendingOrder = false;
	}

	public async createOrderBookSide(
		relayerClient: RelayerClient,
		pair: string,
		bestPrice: number,
		isBid: boolean,
		level: number,
		adjustPrice: boolean = true
	) {
		const precision = this.tokens[0].precisions[Constants.TOKEN_WETH];
		this.isSendingOrder = true;
		for (let i = 0; i < level; i++) {
			const levelPrice = Number(
				Util.formatFixedNumber(
					bestPrice +
						(adjustPrice
							? (isBid ? -1 : 1) *
							(i + CST.MIN_ORDER_BOOK_LEVELS - level) *
							this.priceStep
							: 0),
					precision
				)
			);
			const orderHash = await relayerClient.addOrder(
				this.makerAccount.address,
				pair,
				levelPrice,
				20 + Number((Math.random() * 5).toFixed(1)),
				isBid,
				Util.getExpiryTimestamp(true)
			);
			this.pendingOrders[orderHash] = true;
			await Util.sleep(1000);
		}
		this.isSendingOrder = false;
	}

	public async cancelOrders(relayerClient: RelayerClient, pair: string, orderHashes: string[]) {
		this.isSendingOrder = true;
		orderHashes.forEach(o => (this.pendingOrders[o] = true));
		const signature = await relayerClient.web3Util.web3PersonalSign(
			this.makerAccount.address,
			Constants.TERMINATE_SIGN_MSG + orderHashes.join(',')
		);
		relayerClient.deleteOrder(pair, orderHashes, signature);
		this.isSendingOrder = false;
	}

	public handleOrderError(method: string, orderHash: string, error: string) {
		Util.logError(method + ' ' + orderHash + ' ' + error);
		if (this.pendingOrders[orderHash]) delete this.pendingOrders[orderHash];
		// TODO: handle add and terminate error
	}

	public getDualWrapper(
		contractType: string,
		infuraProvider: string,
		aToken: IToken,
		live: boolean
	) {
		const web3Wrapper = new Web3Wrapper(
			null,
			infuraProvider,
			this.makerAccount.privateKey,
			live
		);

		const dualClassWrapper =
			contractType === WrapperConstants.VIVALDI
				? new VivaldiWrapper(web3Wrapper, aToken.custodian)
				: new DualClassWrapper(web3Wrapper, aToken.custodian);
		return dualClassWrapper;
	}

	public async initialize(relayerClient: RelayerClient, option: IOption) {
		Util.logInfo('initializing dual class wrapper');
		this.isSendingOrder = false;
		this.isMaintainingBalance = false;
		this.isMakingOrders = false;
		const live = option.env === Constants.DB_LIVE;
		const aToken = relayerClient.web3Util.getTokenByCode(option.token);
		if (!aToken) throw new Error('no aToken');
		const bToken = relayerClient.web3Util.tokens.find(
			t => t.code !== aToken.code && t.custodian === aToken.custodian
		);
		if (!bToken) throw new Error('no bToken');
		this.tokens = [aToken, bToken];
		this.priceStep = aToken.precisions[Constants.TOKEN_WETH] * 20;
		let infura = {
			token: ''
		};
		try {
			infura = require('../keys/infura.json');
		} catch (error) {
			Util.logError(`error in loading infura token: ${JSON.stringify(error)}`);
		}
		const infuraProvider =
			(live ? Constants.PROVIDER_INFURA_MAIN : Constants.PROVIDER_INFURA_KOVAN) +
			'/' +
			infura.token;
		const dualWrapper = this.getDualWrapper(option.contractType, infuraProvider, aToken, live);
		const address = this.makerAccount.address;
		Util.logDebug(`updating balance for maker address ${address}`);
		this.tokenBalances = [
			await relayerClient.web3Util.getTokenBalance(Constants.TOKEN_WETH, address),
			await relayerClient.web3Util.getTokenBalance(this.tokens[0].code, address),
			await relayerClient.web3Util.getTokenBalance(this.tokens[1].code, address)
		];
		Util.logDebug('token balances: ' + JSON.stringify(this.tokenBalances));
		await this.checkAllowance(relayerClient.web3Util, dualWrapper);
		await this.maintainBalance(relayerClient.web3Util, dualWrapper as DualClassWrapper);
		relayerClient.subscribeOrderHistory(this.makerAccount.address);
		return dualWrapper;
	}
}
