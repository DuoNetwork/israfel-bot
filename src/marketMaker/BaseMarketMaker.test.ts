// fix for @ledgerhq/hw-transport-u2f 4.28.0
import '@babel/polyfill';
// import { DualClassWrapper } from '@finbook/duo-contract-wrapper';
import { Constants as DataConstants } from '@finbook/duo-market-data';
import { Util } from '@finbook/israfel-common';
import BaseMarketMaker from './BaseMarketMaker';

const baseMarketMaker = new BaseMarketMaker();

// const userOrders = [
	// {
	// 	account: 'account',
	// 	pair: 'aETH|WETH',
	// 	orderHash: 'orderHash1',
	// 	price: 0.001,
	// 	amount: 10,
	// 	balance: 10,
	// 	matching: 0,
	// 	fill: 0,
	// 	side: 'bid',
	// 	expiry: 1234567890000,
	// 	createdAt: 1234567880000,
	// 	updatedAt: 1234567880000,
	// 	initialSequence: 1,
	// 	currentSequence: 1,
	// 	fee: 0.1,
	// 	feeAsset: 'aETH',
	// 	type: 'add',
	// 	status: 'confirmed',
	// 	updatedBy: 'relayer',
	// 	processed: true,
	// 	transactionHash: 'transactionhash1'
	// },
	// {
	// 	account: 'account',
	// 	pair: 'bETH|WETH',
	// 	orderHash: 'orderHash2',
	// 	price: 0.0011,
	// 	amount: 10,
	// 	balance: 12,
	// 	matching: 0,
	// 	fill: 0,
	// 	side: 'bid',
	// 	expiry: 1234567890000,
	// 	createdAt: 1234567880000,
	// 	updatedAt: 1234567880000,
	// 	initialSequence: 2,
	// 	currentSequence: 2,
	// 	fee: 0.1,
	// 	feeAsset: 'bETH',
	// 	type: 'add',
	// 	status: 'confirmed',
	// 	updatedBy: 'relayer',
	// 	processed: true,
	// 	transactionHash: 'transactionhash2'
	// },
	// {
	// 	account: 'account',
	// 	pair: 'aETH|WETH',
	// 	orderHash: 'orderHash3',
	// 	price: 0.0013,
	// 	amount: 10,
	// 	balance: 14,
	// 	matching: 0,
	// 	fill: 0,
	// 	side: 'ask',
	// 	expiry: 1234567890000,
	// 	createdAt: 1234567880000,
	// 	updatedAt: 1234567880000,
	// 	initialSequence: 3,
	// 	currentSequence: 3,
	// 	fee: 0.1,
	// 	feeAsset: 'aETH',
	// 	type: 'add',
	// 	status: 'confirmed',
	// 	updatedBy: 'relayer',
	// 	processed: true,
	// 	transactionHash: 'transactionhash3'
	// },
	// {
	// 	account: 'account',
	// 	pair: 'bETH|WETH',
	// 	orderHash: 'orderHash4',
	// 	price: 0.0014,
	// 	amount: 10,
	// 	balance: 16,
	// 	matching: 0,
	// 	fill: 0,
	// 	side: 'ask',
	// 	expiry: 1234567890000,
	// 	createdAt: 1234567880000,
	// 	updatedAt: 1234567880000,
	// 	initialSequence: 4,
	// 	currentSequence: 4,
	// 	fee: 0.1,
	// 	feeAsset: 'aETH',
	// 	type: 'add',
	// 	status: 'confirmed',
	// 	updatedBy: 'relayer',
	// 	processed: true,
	// 	transactionHash: 'transactionhash4'
	// },
	// {
	// 	account: 'account',
	// 	pair: 'aETH|WETH',
	// 	orderHash: 'orderHash1',
	// 	price: 0.001,
	// 	amount: 10,
	// 	balance: 10,
	// 	matching: 0,
	// 	fill: 0,
	// 	side: 'bid',
	// 	expiry: 1234567890000,
	// 	createdAt: 1234567880000,
	// 	updatedAt: 1234567880000,
	// 	initialSequence: 1,
	// 	currentSequence: 5,
	// 	fee: 0.1,
	// 	feeAsset: 'aETH',
	// 	type: 'terminate',
	// 	status: 'confirmed',
	// 	updatedBy: 'relayer',
	// 	processed: true,
	// 	transactionHash: 'transactionhash1'
// 	}
// ];

const tokens = [
	{
		custodian: '0x56e2727e56F9D6717e462418f822a8FE08Be4711',
		address: 'address',
		code: 'aETH',
		denomination: 0.1,
		precisions: {
			WETH: 0.000005
		},
		feeSchedules: {
			WETH: {
				minimum: 0.1,
				rate: 0
			}
		}
	},
	{
		custodian: '0x56e2727e56F9D6717e462418f822a8FE08Be4711',
		address: 'address',
		code: 'bETH',
		denomination: 0.1,
		precisions: {
			WETH: 0.000005
		},
		feeSchedules: {
			WETH: {
				minimum: 0.1,
				rate: 0
			}
		}
	}
];

const option = {
	env: 'dev',
	tokens: [],
	token: 'aETH',
	contractType: 'Beethoven',
	maker: 0,
	spender: 1,
	amount: 10,
	debug: true,
	server: false
};

baseMarketMaker.makerAccount = {
	address: 'address',
	privateKey: 'privateKey'
};
test('checkAllowance, already approved', async () => {
	const web3Util = {
		getTokenAllowance: jest.fn(() => Promise.resolve(10000)),
		setUnlimitedTokenAllowance: jest.fn(() => Promise.resolve()),
		awaitTransactionSuccessAsync: jest.fn(() => Promise.resolve())
	} as any;

	const dualClassWrapper1 = {
		address: 'custodianAddr'
	} as any;
	baseMarketMaker.tokens = tokens;
	await baseMarketMaker.checkAllowance(web3Util, dualClassWrapper1);
	expect((web3Util.getTokenAllowance as jest.Mock).mock.calls).toMatchSnapshot();
	expect(web3Util.setUnlimitedTokenAllowance as jest.Mock).not.toBeCalled();
	expect(web3Util.awaitTransactionSuccessAsync as jest.Mock).not.toBeCalled();
});

test('checkAllowance, 0 allowance', async () => {
	const web3Util = {
		getTokenAllowance: jest.fn(() => Promise.resolve(0)),
		setUnlimitedTokenAllowance: jest.fn((code: string, addr: string, custodianAddr?: string) =>
			Promise.resolve(`${code}|
	${addr}|${custodianAddr}`)
		),
		awaitTransactionSuccessAsync: jest.fn(() => Promise.resolve())
	} as any;

	const dualClassWrapper1 = {
		address: 'custodianAddr'
	} as any;
	baseMarketMaker.tokens = tokens;
	await baseMarketMaker.checkAllowance(web3Util, dualClassWrapper1);
	expect((web3Util.getTokenAllowance as jest.Mock).mock.calls).toMatchSnapshot();
	expect((web3Util.setUnlimitedTokenAllowance as jest.Mock).mock.calls).toMatchSnapshot();
	expect((web3Util.awaitTransactionSuccessAsync as jest.Mock).mock.calls).toMatchSnapshot();
});

test('initialize, no a token', async () => {
	const web3Util = {
		getTokenByCode: jest.fn(() => null),
		tokens: [],
		getTokenBalance: jest.fn(() => Promise.resolve(10))
	} as any;
	const relayerClient = {
		web3Util: web3Util,
		subscribeOrderBook: jest.fn(() => Promise.resolve())
	} as any;

	try {
		await baseMarketMaker.initialize(relayerClient, option);
		expect(false).toBeTruthy();
	} catch (err) {
		expect(err).toMatchSnapshot();
	}
});

test('initialize, no b token', async () => {
	const web3Util = {
		getTokenByCode: jest.fn(() => 'aETH'),
		tokens: [tokens[0]],
		getTokenBalance: jest.fn(() => Promise.resolve(10))
	} as any;
	const relayerClient = {
		web3Util: web3Util,
		subscribeOrderBook: jest.fn(() => Promise.resolve())
	} as any;

	try {
		await baseMarketMaker.initialize(relayerClient, option);
		expect(false).toBeTruthy();
	} catch (err) {
		expect(err).toMatchSnapshot();
	}
});

test('initialize', async () => {
	const web3Util = {
		getTokenByCode: jest.fn(() => tokens[0]),
		tokens: tokens,
		getTokenBalance: jest.fn(() => Promise.resolve(10))
	} as any;
	const relayerClient = {
		web3Util: web3Util,
		subscribeOrderHistory: jest.fn(() => Promise.resolve())
	} as any;

	baseMarketMaker.checkAllowance = jest.fn(() => Promise.resolve());
	baseMarketMaker.maintainBalance = jest.fn(() => Promise.resolve());
	baseMarketMaker.getDualWrapper = jest.fn();

	await baseMarketMaker.initialize(relayerClient, option);
	expect((relayerClient.web3Util.getTokenByCode as jest.Mock).mock.calls).toMatchSnapshot();
	expect((relayerClient.web3Util.getTokenBalance as jest.Mock).mock.calls).toMatchSnapshot();
	expect((relayerClient.subscribeOrderHistory as jest.Mock).mock.calls).toMatchSnapshot();
});

test('cancelOrders', async () => {
	const relayerClient = {
		web3Util: {
			web3PersonalSign: jest.fn(() => Promise.resolve('signature'))
		},
		deleteOrder: jest.fn(() => Promise.resolve())
	} as any;
	await baseMarketMaker.cancelOrders(relayerClient, 'aETH|WETH', ['orderHash1', 'orderHash2']);
	expect((relayerClient.web3Util.web3PersonalSign as jest.Mock).mock.calls).toMatchSnapshot();
	expect((relayerClient.deleteOrder as jest.Mock).mock.calls).toMatchSnapshot();
	expect(baseMarketMaker.pendingOrders).toMatchSnapshot();
});

test('canMakeOrder, no orderBookSnapshot', () => {
	baseMarketMaker.tokens = tokens;
	const relayerClient = {
		orderBookSnapshots: {}
	} as any;
	expect(baseMarketMaker.canMakeOrder(relayerClient, 'aETH|WETH')).toBeFalsy();
});

test('canMakeOrder, isSendingOrder', () => {
	baseMarketMaker.tokens = tokens;
	baseMarketMaker.isSendingOrder = true;
	const relayerClient = {
		orderBookSnapshots: {
			'aETH|WETH': {
				version: 1,
				pair: 'aETH|WETH',
				bids: [],
				asks: []
			},
			'bETH|WETH': { version: 1, pair: 'bETH|WETH', bids: [], asks: [] }
		}
	} as any;
	expect(baseMarketMaker.canMakeOrder(relayerClient, 'bETH|WETH')).toBeFalsy();
});

test('canMakeOrder, has pendingOrder', () => {
	baseMarketMaker.tokens = tokens;
	baseMarketMaker.isSendingOrder = true;
	baseMarketMaker.pendingOrders = { orderHash: true };
	const relayerClient = {
		orderBookSnapshots: {
			'aETH|WETH': {
				version: 1,
				pair: 'aETH|WETH',
				bids: [],
				asks: []
			},
			'bETH|WETH': { version: 1, pair: 'bETH|WETH', bids: [], asks: [] }
		}
	} as any;
	expect(baseMarketMaker.canMakeOrder(relayerClient, 'aETH|WETH')).toBeFalsy();
});

test('canMakeOrder', () => {
	baseMarketMaker.tokens = tokens;
	baseMarketMaker.isSendingOrder = false;
	baseMarketMaker.pendingOrders = {};
	const relayerClient = {
		orderBookSnapshots: {
			'aETH|WETH': {
				version: 1,
				pair: 'aETH|WETH',
				bids: [],
				asks: []
			},
			'bETH|WETH': { version: 1, pair: 'bETH|WETH', bids: [], asks: [] }
		}
	} as any;
	expect(baseMarketMaker.canMakeOrder(relayerClient, 'aETH|WETH')).toBeTruthy();
});

test('getEthPrice', () => {
	baseMarketMaker.exchangePrices[DataConstants.API_KRAKEN] = [
		{
			period: 1,
			open: 100,
			high: 200,
			low: 50,
			close: 150,
			volume: 10000,
			source: 'kraken',
			base: 'USD',
			quote: 'ETH',
			timestamp: 1234567890000
		}
	];
	expect(baseMarketMaker.getEthPrice()).toBe(150);
});

test('getEthPrice, no ETH price', () => {
	baseMarketMaker.exchangePrices[DataConstants.API_KRAKEN] = [];
	expect(baseMarketMaker.getEthPrice()).toBe(0);
});

test('takeOneSideOrders bid', async () => {
	const relayerClient = {
		addOrder: jest.fn(() => Promise.resolve('addOrderTxHash'))
	} as any;
	Util.getExpiryTimestamp = jest.fn(() => 1234567890000);
	Util.sleep = jest.fn(() => Promise.resolve()) as any;
	baseMarketMaker.pendingOrders = {};
	await baseMarketMaker.takeOneSideOrders(relayerClient, 'aETH|WETH', true, [
		{
			price: 0.001,
			balance: 20,
			count: 1
		},
		{
			price: 0.0012,
			balance: 0,
			count: 1
		},
		{
			price: 0.0014,
			balance: 20,
			count: 1
		}
	]);
	expect((relayerClient.addOrder as jest.Mock).mock.calls).toMatchSnapshot();
	expect(baseMarketMaker.pendingOrders).toMatchSnapshot();
});

test('takeOneSideOrders ask', async () => {
	const relayerClient = {
		addOrder: jest.fn(() => Promise.resolve('addOrderTxHash'))
	} as any;
	Util.getExpiryTimestamp = jest.fn(() => 1234567890000);
	Util.sleep = jest.fn(() => Promise.resolve()) as any;
	baseMarketMaker.pendingOrders = {};
	await baseMarketMaker.takeOneSideOrders(relayerClient, 'aETH|WETH', false, [
		{
			price: 0.001,
			balance: 20,
			count: 1
		},
		{
			price: 0.0012,
			balance: 0,
			count: 1
		},
		{
			price: 0.0014,
			balance: 20,
			count: 1
		}
	]);
	expect((relayerClient.addOrder as jest.Mock).mock.calls).toMatchSnapshot();
	expect(baseMarketMaker.pendingOrders).toMatchSnapshot();
});

test('createOrderBookSide bid', async () => {
	const relayerClient = {
		addOrder: jest.fn(() => Promise.resolve('addOrderTxHash'))
	} as any;
	Util.getExpiryTimestamp = jest.fn(() => 1234567890000);
	Util.sleep = jest.fn(() => Promise.resolve()) as any;
	Math.random = jest.fn(() => 0.5);
	baseMarketMaker.pendingOrders = {};
	await baseMarketMaker.createOrderBookSide(relayerClient, 'aETH|WETH', 0.0001, true, 4);
	expect((relayerClient.addOrder as jest.Mock).mock.calls).toMatchSnapshot();
	expect(baseMarketMaker.pendingOrders).toMatchSnapshot();
});

test('createOrderBookSide bid, no price adjustment', async () => {
	const relayerClient = {
		addOrder: jest.fn(() => Promise.resolve('addOrderTxHash'))
	} as any;
	Util.getExpiryTimestamp = jest.fn(() => 1234567890000);
	Util.sleep = jest.fn(() => Promise.resolve()) as any;
	Math.random = jest.fn(() => 0.5);
	baseMarketMaker.pendingOrders = {};
	await baseMarketMaker.createOrderBookSide(relayerClient, 'aETH|WETH', 0.0001, true, 4, false);
	expect((relayerClient.addOrder as jest.Mock).mock.calls).toMatchSnapshot();
	expect(baseMarketMaker.pendingOrders).toMatchSnapshot();
});

test('createOrderBookSide ask', async () => {
	const relayerClient = {
		addOrder: jest.fn(() => Promise.resolve('addOrderTxHash'))
	} as any;
	Util.getExpiryTimestamp = jest.fn(() => 1234567890000);
	Util.sleep = jest.fn(() => Promise.resolve()) as any;
	Math.random = jest.fn(() => 0.5);
	baseMarketMaker.pendingOrders = {};
	await baseMarketMaker.createOrderBookSide(relayerClient, 'aETH|WETH', 0.0001, false, 4);
	expect((relayerClient.addOrder as jest.Mock).mock.calls).toMatchSnapshot();
	expect(baseMarketMaker.pendingOrders).toMatchSnapshot();
});
