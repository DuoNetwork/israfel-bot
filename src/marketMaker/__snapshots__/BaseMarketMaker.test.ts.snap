// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cancelOrders 1`] = `
Array [
  Array [
    "address",
    "By signing this message, your are confirming to cancel following orders (no gas cost involved): orderHash1,orderHash2",
  ],
]
`;

exports[`cancelOrders 2`] = `
Array [
  Array [
    "aETH|WETH",
    Array [
      "orderHash1",
      "orderHash2",
    ],
    "signature",
  ],
]
`;

exports[`cancelOrders 3`] = `
Object {
  "orderHash1": true,
  "orderHash2": true,
}
`;

exports[`checkAllowance, 0 allowance 1`] = `
Array [
  Array [
    "WETH",
    "address",
  ],
  Array [
    "aETH",
    "address",
  ],
  Array [
    "bETH",
    "address",
  ],
  Array [
    "WETH",
    "address",
    "custodianAddr",
  ],
]
`;

exports[`checkAllowance, 0 allowance 2`] = `
Array [
  Array [
    "WETH",
    "address",
  ],
  Array [
    "aETH",
    "address",
  ],
  Array [
    "bETH",
    "address",
  ],
  Array [
    "WETH",
    "address",
    "custodianAddr",
  ],
]
`;

exports[`checkAllowance, 0 allowance 3`] = `
Array [
  Array [
    "WETH|
	address|undefined",
  ],
  Array [
    "aETH|
	address|undefined",
  ],
  Array [
    "bETH|
	address|undefined",
  ],
  Array [
    "WETH|
	address|custodianAddr",
  ],
]
`;

exports[`checkAllowance, already approved 1`] = `
Array [
  Array [
    "WETH",
    "address",
  ],
  Array [
    "aETH",
    "address",
  ],
  Array [
    "bETH",
    "address",
  ],
  Array [
    "WETH",
    "address",
    "custodianAddr",
  ],
]
`;

exports[`createOrderBookSide ask 1`] = `
Array [
  Array [
    "address",
    "aETH|WETH",
    0.0001,
    22.5,
    false,
    1234567890000,
  ],
  Array [
    "address",
    "aETH|WETH",
    0.0002,
    22.5,
    false,
    1234567890000,
  ],
  Array [
    "address",
    "aETH|WETH",
    0.0003,
    22.5,
    false,
    1234567890000,
  ],
  Array [
    "address",
    "aETH|WETH",
    0.0004,
    22.5,
    false,
    1234567890000,
  ],
]
`;

exports[`createOrderBookSide ask 2`] = `
Object {
  "addOrderTxHash": true,
}
`;

exports[`createOrderBookSide bid 1`] = `
Array [
  Array [
    "address",
    "aETH|WETH",
    0.0001,
    22.5,
    true,
    1234567890000,
  ],
  Array [
    "address",
    "aETH|WETH",
    0,
    22.5,
    true,
    1234567890000,
  ],
  Array [
    "address",
    "aETH|WETH",
    -0.0001,
    22.5,
    true,
    1234567890000,
  ],
  Array [
    "address",
    "aETH|WETH",
    -0.0002,
    22.5,
    true,
    1234567890000,
  ],
]
`;

exports[`createOrderBookSide bid 2`] = `
Object {
  "addOrderTxHash": true,
}
`;

exports[`createOrderBookSide bid, no price adjustment 1`] = `
Array [
  Array [
    "address",
    "aETH|WETH",
    0.0001,
    22.5,
    true,
    1234567890000,
  ],
  Array [
    "address",
    "aETH|WETH",
    0.0001,
    22.5,
    true,
    1234567890000,
  ],
  Array [
    "address",
    "aETH|WETH",
    0.0001,
    22.5,
    true,
    1234567890000,
  ],
  Array [
    "address",
    "aETH|WETH",
    0.0001,
    22.5,
    true,
    1234567890000,
  ],
]
`;

exports[`createOrderBookSide bid, no price adjustment 2`] = `
Object {
  "addOrderTxHash": true,
}
`;

exports[`initialize 1`] = `
Array [
  Array [
    "aETH",
  ],
]
`;

exports[`initialize 2`] = `
Array [
  Array [
    "WETH",
    "address",
  ],
  Array [
    "aETH",
    "address",
  ],
  Array [
    "bETH",
    "address",
  ],
]
`;

exports[`initialize 3`] = `
Array [
  Array [
    "address",
  ],
]
`;

exports[`initialize, no a token 1`] = `[Error: no aToken]`;

exports[`initialize, no b token 1`] = `[Error: no bToken]`;

exports[`takeOneSideOrders ask 1`] = `
Array [
  Array [
    "address",
    "aETH|WETH",
    0.001,
    20,
    true,
    1234567890000,
  ],
  Array [
    "address",
    "aETH|WETH",
    0.0014,
    20,
    true,
    1234567890000,
  ],
]
`;

exports[`takeOneSideOrders ask 2`] = `
Object {
  "addOrderTxHash": true,
}
`;

exports[`takeOneSideOrders bid 1`] = `
Array [
  Array [
    "address",
    "aETH|WETH",
    0.001,
    20,
    false,
    1234567890000,
  ],
  Array [
    "address",
    "aETH|WETH",
    0.0014,
    20,
    false,
    1234567890000,
  ],
]
`;

exports[`takeOneSideOrders bid 2`] = `
Object {
  "addOrderTxHash": true,
}
`;
